### PR TITLE
first draft

### PR DESCRIPTION
**Pull request by xz62px**

### CODE REVIEW

File: src/zenohSessionManager.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	static ZenohSessionManager sessionManager;
+
+	return sessionManager;
+}
+
+UCode ZenohSessionManager::init(ZenohSessionManagerConfig &sessionConfig) noexcept {
+
+    if (0 == refCount_) {
+        ++refCount_;
+
```
Comment (MelamudMichael) [inline src/zenohSessionManager.cpp:15]: add mutex

File: src/zenohSessionManager.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    spdlog::info("ZenohSessionManager::init done , ref count {}", 
+                 refCount_);
+    
+    return UCode::OK;
+}
+
+UCode ZenohSessionManager::term() noexcept {
+    
+    if (0 < refCount_) {
+
```
Comment (MelamudMichael) [inline src/zenohSessionManager.cpp:53]: add mutex

**Comment by stevenhartley** [comment]
@MelamudMichael don't forget the required Apache header files along with the required legal documents (CONTRIBUTING.adoc, NOTIC.adoc, LICENSE)

File: src/zenohUTransport.cpp [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+
+        if (nullptr == listenerContainer) {
+            listenerContainer = make_shared<ListenerContainer>();
+            if (nullptr == listenerContainer) {
+                spdlog::error("listenerContainer allocation failure");
+                break;
+            }
+        }
+        
+        cbArgumentType *arg = new cbArgumentType(std::move(uri),
```
Comment (linsonittoop) [inline src/zenohUTransport.cpp:370]: Avoid Duplicate definition of cbArgumentType with same name _arg_

Comment (linsonittoop) [inline src/zenohUTransport.cpp:370]: Is the _arg_ pointer stored and deallocated properly?

**Review by linsonittoop — CHANGES_REQUESTED** [review]
Comments from a  partial review.

File: src/messageParser.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        int value = *reinterpret_cast<int*>(map[Tag::PLEVEL].value.data());
+        attributesBuilder.withPermissionLevel(value);
+    }
+
+    if (map.find(Tag::COMMSTATUS) != map.end()) {
+        int value = *reinterpret_cast<int*>(map[Tag::COMMSTATUS].value.data());
+        attributesBuilder.withCommStatus(value);
+    }
+
+    // if (map.find(Tag::TOKEN) != map.end()) {
```
[the remark this answers is not in the document]
Comment (MelamudMichael) [reply src/messageParser.cpp:139]: i will uncomment

File: src/messageParser.cpp [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    if (map.find(Tag::PAYLOAD) == map.end()) {
+        spdlog::error("PAYLOAD tag not found");
+        return std::nullopt;
+    }
+
+    TLV tlvPayload = map[Tag::PAYLOAD];
+
+    spdlog::debug("PAYLOAD tag found length = {}", tlvPayload.length);
+
+    return UPayload(tlvPayload.value.data(),
```
[the remark this answers is not in the document]
Comment (MelamudMichael) [reply src/messageParser.cpp:166]: need to align 1.5 branch

File: src/zenohRpcClient.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	return rpcClient;
+}
+
+UCode ZenohRpcClient::init() noexcept {
+    if (true == init_) {
+        spdlog::error("ZenohRpcClient is already initialized");
+        return UCode::UNAVAILABLE;
+    }
+
+    ZenohSessionManagerConfig config;
```
Comment (MelamudMichael) [inline src/zenohRpcClient.cpp:32]: initialize strings to null

File: src/zenohUTransport.cpp [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+            }
+        } else {
+            options.encoding = z_encoding(Z_ENCODING_PREFIX_TEXT_PLAIN, 
+                                          nullptr);
+        }
+
+        auto message = MessageBuilder::instance().build(uri,
+                                                        attributes,
+                                                        payload);
+        if (true == message.size()) {
```
[the remark this answers is not in the document]
Comment (MelamudMichael) [reply src/zenohUTransport.cpp:227]: should be 0

File: src/zenohUTransport.cpp [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+    }
+
+    init_ = false;
+
+    spdlog::info("ZenohUTransport term done");
+    
+    return UCode::OK;
+}
+
+UCode ZenohUTransport::authenticate(const UEntity &uEntity)
```
[the remark this answers is not in the document]
Comment (MelamudMichael) [reply src/zenohUTransport.cpp:113]: it is not implemented yet

File: src/messageBuilder.cpp [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    spdlog::debug("total message size = {}", 
+                  messageSize);
+
+    std::vector<uint8_t> message(messageSize);
+
+    size_t pos = 0;
+
+    auto uriBase64 = uprotocol::tools::Base64::encode(uri.toString());
+
+    // pos = addTag(message,
```
[the remark this answers is not in the document]
Comment (MelamudMichael) [reply src/messageBuilder.cpp:79]: no

**Review by linsonittoop — CHANGES_REQUESTED** [review]

**Comment by stevenhartley** [comment]
This was created to the wrong project

**Pull request closed without merging**